elasticdl/python/tests/optimizer_wrapper_test.py

```diff
@@ -620,6 +625,188 @@ def test_correctness(self):
+        embed_var_n = len(embed_values)
+        mock_kv_store = MockKvStore()
+        for layer, values in embed_values.items():
+            embed_dims[layer] = values.shape[1]
+            input_dim = values.shape[0]
+
+            keys = [
+                Embedding.get_key([layer, idx]) for idx in range(input_dim)
+            ]
+            mock_kv_store.update(keys, values)
+
+        opt = SGD(0.1)
+        opt_wrapper = OptimizerWrapper(opt, None, embed_dims, True)
+
+        with mock.patch.object(
+            EmbeddingService, "lookup_embedding", mock_kv_store.lookup
+        ), mock.patch.object(
+            EmbeddingService, "update_embedding", mock_kv_store.update
+        ):
+            # call optimizer_wrapper.apply_gradients asynchronously
+            def _apply_gradients(opt_wrapper, grads_and_vars):
+                # sleep 1s to wait that all threads are in this method call
+                time.sleep(1)
+                opt_wrapper.apply_gradients(grads_and_vars)
+
+            executor = ThreadPoolExecutor(max_workers=thread_num)
+            tasks = [
+                executor.submit(_apply_gradients, opt_wrapper, grads_and_vars)
+                for grads_and_vars in grads_and_vars_batches
+            ]
+            _ = [task.result() for task in tasks]
+
+            # check updated results of non-embedding variables
+            non_embed_vars = [
+                var for grad, var in grads_and_vars_batches[0][:-embed_var_n]
+            ]
+            for var, expected_value in zip(
+                non_embed_vars, expected_non_embed_values
+            ):
+                self.assertTrue(np.isclose(var.numpy(), expected_value).all())
+
+            # `expected_embed_values=None` means that no need to check
+            # embedding table
+            if not expected_embed_values:
+                return
+            # check updated results of embedding table
+            for layer, expected_values in expected_embed_values.items():
+                keys = [
+                    Embedding.get_key([layer, idx]) for idx in range(input_dim)
+                ]
+                raw_value, _ = mock_kv_store.lookup(keys)
+                value = np.concatenate(raw_value).reshape(input_dim, -1)
+
+                self.assertTrue(
+                    any(
+                        [
+                            np.isclose(value, expected).all()
+                            for expected in expected_values
+                        ]
+                    )
+                )
+
+    def test_async_correctness(self):
+        """Tests the correctness of async updates in `OptimizerWrapper`.
+
+        Testing the correctness is not simple because OptimizerWrapper is not
+        thread-safe for embedding table. This test case lists all the possible
+        results when `thread_number=2` and test the correctness. This test case
+        also tests that OptimizerWrapper does not raise Error when
+        `thread_num=8.`
```

Comment:
Why 8?

Reply:
7, 8, 9, 10 are both ok. I just choose a big number for `thread_num`. Because When the number of threads is small, there may be bugs in the code, but the probability of bugs occurring is small.

Reply:
Perhaps "when `thread_num=8`" -> "with bigger thread number"?